test/common-tests/acbs-authentication-api-tests.ts

```diff
@@ -0,0 +1,134 @@
+import { RandomValueGenerator } from '@ukef-test/support/generator/random-value-generator';
+import nock from 'nock';
+import request from 'supertest';
+
+interface AcbsAuthenticationErrorCasesTestOptions {
+  givenRequestWouldOtherwiseSucceed: () => void;
+  makeRequest: () => request.Test;
+}
+
+interface AcbsAuthenticationTestHooks {
+  idToken: string;
+  givenAuthenticationWithTheIdpSucceeds: () => void;
+}
+
+// Note that these tests assume the caller will use nock to abort pending requests
+// and clean all mocks after each test
+export const withAcbsAuthenticationApiTests = ({
+  givenRequestWouldOtherwiseSucceed,
+  makeRequest,
+}: AcbsAuthenticationErrorCasesTestOptions): AcbsAuthenticationTestHooks => {
+  const valueGenerator = new RandomValueGenerator();
+  const idToken = valueGenerator.string();
+
+  it('returns a 500 response if creating a session with the IdP fails', async () => {
+    const errorCode = valueGenerator.string();
+    requestToCreateASessionWithTheIdp().reply(
+      400,
+      {
+        errorCode: errorCode,
+        errorStack: null,
+        messages: [
+          {
+            code: errorCode,
+            message: `${errorCode}: Invalid Credentials. (msg_id=${valueGenerator.string()})`,
+            property: null,
+          },
+        ],
+      },
+      {
+        'set-cookie': 'JSESSIONID=prelogin-1',
+      },
+    );
+    givenRequestWouldOtherwiseSucceed();
+
+    const { status, body } = await makeRequest();
+
+    expect(status).toBe(500);
+    expect(body).toStrictEqual({
+      statusCode: 500,
+      message: 'Internal server error',
+    });
+  });
+
+  it('returns a 500 response if getting an id token from the IdP fails', async () => {
+    givenCreatingASessionWithTheIdpSucceeds();
+    requestToGetAnIdTokenFromTheIdp().reply(403, '<!doctype html><html><body><div>Access to the requested resource has been forbidden.</div></body></html>');
+    givenRequestWouldOtherwiseSucceed();
+
+    const { status, body } = await makeRequest();
+
+    expect(status).toBe(500);
+    expect(body).toStrictEqual({
+      statusCode: 500,
+      message: 'Internal server error',
+    });
+  });
+
+  it('returns a 500 response if getting an id token from the IdP times out', async () => {
+    givenCreatingASessionWithTheIdpSucceeds();
+    requestToGetAnIdTokenFromTheIdp()
+      .delay(ENVIRONMENT_VARIABLES.ACBS_TIMEOUT + 500)
+      .reply(200, { id_token: idToken });
+    givenRequestWouldOtherwiseSucceed();
+
+    const { status, body } = await makeRequest();
+
+    expect(status).toBe(500);
+    expect(body).toStrictEqual({
+      statusCode: 500,
+      message: 'Internal server error',
+    });
+  });
+
+  it('returns a 500 response if creating a session with the IdP times out', async () => {
+    requestToCreateASessionWithTheIdp()
+      .delay(ENVIRONMENT_VARIABLES.ACBS_TIMEOUT + 500)
+      .reply(201, '', { 'set-cookie': 'JSESSIONID=1' });
```

Comment:
Can we please add cookie name in `@ukef/constants`, as during security improvement default cookie name will be changed. Thanks

Reply:
I've added this to `@ukef/constants` - let me know if it wasn't exactly what you had in mind.

As far as I'm aware, the name of this cookie is controlled by the 3rd party API we authenticate with, rather than by us. Are we expecting the 3rd party to change the name of the cookie in the future?